runtime/resetting_histogram.go

```diff
@@ -0,0 +1,298 @@
+package runtime
+
+import (
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/rcrowley/go-metrics"
+)
+
+// Initial slice capacity for the values stored in a ResettingHistogram
+const InitialResettingHistogramSliceCap = 10
+
+// NewResettingHistogram constructs a new StandardResettingHistogram
+func NewResettingHistogram() Histogram {
+	if metrics.UseNilMetrics {
+		return NilResettingHistogram{}
+	}
+	return &StandardResettingHistogram{
+		values: make([]int64, 0, InitialResettingHistogramSliceCap),
+	}
+}
+
+// NilResettingHistogram is a no-op ResettingHistogram.
+type NilResettingHistogram struct {
+}
+
+// Values is a no-op.
+func (NilResettingHistogram) Values() []int64 { return nil }
+
+// Snapshot is a no-op.
+func (NilResettingHistogram) Snapshot() Histogram {
+	return &ResettingHistogramSnapshot{
+		values: []int64{},
+	}
+}
+
+// Update is a no-op.
+func (NilResettingHistogram) Update(int64) {}
+
+// Clear is a no-op.
+func (NilResettingHistogram) Clear() {}
+
+func (NilResettingHistogram) Count() int64 {
+	panic("Count called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Variance() float64 {
+	panic("Count called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Min() int64 {
+	panic("Min called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Max() int64 {
+	panic("Max called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Sum() int64 {
+	panic("Sum called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) StdDev() float64 {
+	panic("StdDev called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Sample() Sample {
+	panic("Sample called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Percentiles([]float64) []float64 {
+	panic("Percentiles called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Percentile(float64) float64 {
+	panic("Percentiles called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Mean() float64 {
+	panic("Mean called on a NilResettingHistogram")
+}
+
+// StandardResettingHistogram is used for storing aggregated values for timers, which are reset on every flush interval.
+type StandardResettingHistogram struct {
+	values []int64
+	mutex  sync.Mutex
+}
+
+func (t *StandardResettingHistogram) Count() int64 {
+	panic("Count called on a StandardResettingHistogram")
```

Comment:
It would be terrible to recalculate on every call to Min, Max, Variance, etc. if you call them directly on the Histogram. The idea is to call a Snapshot every N-seconds (interval), and calculate once.

The panics signal to the user that if they call Histogram.Min they are doing it wrong.

Reply:
Yeah, I agree, but the upstream library does support this calculation and I think we should do too. I don't see any reason to drift?

Reply:
The upstream library is operating on samples, this data structure is holding all values and mostly calculating percentiles on them.

If we return these calculations on the `StandardResettingHistogram`, then what is the point of having a `Snapshot`. Also we'd be locking the `Histogram`, and blocking any new updates, while processing the current values.

I really don't see a point in supporting these function on the `Histogram` level.

Reply:
Also you just asked me to unexport all these types - nobody, but our `runtime` package code can actually use these. We are not building a library that is extending the `go-metrics` apparently, that other folks can use. Our code is using these data structures correctly, by generating a snapshot on every interval prior to emitting the measurements.